engine/collection/epochmgr/engine.go

```diff
@@ -207,32 +231,53 @@ func (e *Engine) onEpochTransition(first *flow.Header) error {
 
 	log.Info().Msg("epoch transition: new epoch components started successfully")
 
-	// Finally we set up a trigger to shut down the components of the previous
-	// epoch once we can no longer produce valid collections there. Transactions
-	// referencing blocks from the previous epoch are only valid for inclusion
-	// in collections built by clusters from that epoch. Consequently, it remains
-	// possible for the previous epoch's cluster to produce valid collections
-	// until all such transactions have expired. In fact, since these transactions
-	// can NOT be included by clusters in the new epoch, we MUST continue producing
-	// these collections within the previous epoch's clusters.
-	e.heightEvents.OnHeight(first.Height+flow.DefaultTransactionExpiry, func() {
+	// set up callback to stop previous epoch
+	e.prepareToStopEpochComponents(counter, lastEpochMaxHeight)
+
+	return nil
+}
+
+// prepareToStopEpochComponents registers a callback to stop the epoch with the
+// given counter once it is no longer possible to receive transactions from that
+// epoch. This occurs when we finalize sufficiently many blocks in the new epoch
+// that a transaction referencing any block from the previous epoch would be
+// considered immediately expired.
+//
+// Transactions referencing blocks from the previous epoch are only valid for
+// inclusion in collections built by clusters from that epoch. Consequently, it
+// remains possible for the previous epoch's cluster to produce valid collections
+// until all such transactions have expired. In fact, since these transactions
+// can NOT be included by clusters in the new epoch, we MUST continue producing
+// these collections within the previous epoch's clusters.
+//
+func (e *Engine) prepareToStopEpochComponents(epochCounter, epochMaxHeight uint64) {
+
+	stopAtHeight := epochMaxHeight + flow.DefaultTransactionExpiry + 1
+
+	log := e.log.With().
+		Uint64("epoch_max_height", epochMaxHeight).
+		Uint64("epoch_counter", epochCounter).
+		Uint64("stop_at_height", stopAtHeight).
+		Logger()
+
+	log.Debug().Msgf("preparing to stop epoch components at height %d", stopAtHeight)
+
+	e.heightEvents.OnHeight(stopAtHeight, func() {
 		e.unit.Launch(func() {
 			e.unit.Lock()
 			defer e.unit.Unlock()
 
 			log.Info().Msg("epoch transition: stopping components for previous epoch...")
 
-			err := e.stopEpochComponents(counter - 1)
+			err := e.stopEpochComponents(epochCounter - 1)
 			if err != nil {
-				e.log.Error().Err(err).Msgf("epoch transition: failed to stop components for epoch %d", counter-1)
+				e.log.Error().Err(err).Msgf("epoch transition: failed to stop components for epoch %d", epochCounter-1)
 				return
```

Comment:
As `"epoch transition: "` keeps repeating on the logs, maybe it makes more sense to refactor as 
```
e.log.Error().Err(err).Str("status","epoch transition"). ...
```